lib/value/src/kind/crud/get.rs

```diff
@@ -0,0 +1,474 @@
+            return Self::never();
+        }
+
+        match iter.next() {
+            Some(BorrowedSegment::Field(field) | BorrowedSegment::CoalesceEnd(field)) => {
+                self.get_field(field).get_recursive(iter)
+            }
+            Some(BorrowedSegment::Index(mut index)) => {
+                if let Some(array) = self.as_array() {
+                    if index < 0 {
+                        let largest_known_index = array.known().keys().map(|i| i.to_usize()).max();
+                        // The minimum size of the resulting array.
+                        let len_required = -index as usize;
+
+                        if array.unknown_kind().contains_any_defined() {
+                            // The exact length is not known. We can't know for sure if the index
+                            // will point to a known or unknown type, so the union of the unknown type
+                            // plus any possible known type must be taken. Just the unknown type alone is not sufficient.
+
+                            // The array may be larger, but this is the largest we can prove the array is from the type information.
+                            let min_length = largest_known_index.map_or(0, |i| i + 1);
+
+                            // We can prove the positive index won't be less than "min_index".
+                            let min_index = (min_length as isize + index).max(0) as usize;
+                            let can_underflow = (min_length as isize + index) < 0;
+
+                            let mut kind = array.unknown_kind();
+
+                            // We can prove the index won't underflow, so it cannot be "undefined".
+                            // But only if the type can only be an array.
+                            if self.is_exact() && !can_underflow {
+                                kind.remove_undefined();
+                            }
+
+                            for (i, i_kind) in array.known() {
+                                if i.to_usize() >= min_index {
+                                    kind.merge_keep(i_kind.clone(), false);
+                                }
+                            }
+                            return kind.get_recursive(iter);
+                        }
+
+                        // There are no unknown indices, so we can determine the exact positive index.
+                        let exact_len = largest_known_index.map_or(0, |x| x + 1);
+                        if exact_len >= len_required {
+                            // Make the index positive, then continue below.
+                            index += exact_len as isize;
+                        } else {
+                            // Out of bounds index.
+                            return Self::undefined();
+                        }
+                    }
+
+                    debug_assert!(index >= 0, "negative indices already handled");
```

Comment:
Intended to be included? Two more `debug_assert!()`s in the insert file as well.

Reply:
Yes, those are to verify correctness. These are disabled in release builds.

Reply:
I read the doc page for it, but totally skipped over that section somehow. 👍